loader/loader_linux.c

```diff
@@ -0,0 +1,458 @@
+//   2) Group with the best device type for device 0 wins
+//   3) Group with best PCI bus ID for device 0 wins
+//   4) Ties broken by group device 0 device_ID XOR vendor_ID comparison
+int32_t compare_device_groups(const void *a, const void *b) {
+    struct loader_physical_device_group_term *grp_a = (struct loader_physical_device_group_term *)a;
+    struct loader_physical_device_group_term *grp_b = (struct loader_physical_device_group_term *)b;
+
+    // Use the first GPU's info from each group to sort the groups by
+    struct LinuxSortedDeviceInfo *left = &grp_a->internal_device_info[0];
+    struct LinuxSortedDeviceInfo *right = &grp_b->internal_device_info[0];
+
+    // Default device always gets priority
+    if (left->default_device) {
+        return -1;
+    } else if (right->default_device) {
+        return 1;
+    }
+
+    // Order by device type next
+    int32_t dev_type_comp = device_type_compare(left->device_type, right->device_type);
+    if (0 != dev_type_comp) {
+        return dev_type_comp;
+    }
+
+    // Sort by PCI info (prioritize devices that have info over those that don't)
+    if (left->has_pci_bus_info && !right->has_pci_bus_info) {
+        return -1;
+    } else if (!left->has_pci_bus_info && right->has_pci_bus_info) {
+        return 1;
+    } else if (left->has_pci_bus_info && right->has_pci_bus_info) {
+        // Sort low to high PCI domain
+        if (left->pci_domain < right->pci_domain) {
+            return -1;
+        } else if (left->pci_domain > right->pci_domain) {
+            return 1;
+        }
+        // Sort low to high PCI bus
+        if (left->pci_bus < right->pci_bus) {
+            return -1;
+        } else if (left->pci_bus > right->pci_bus) {
+            return 1;
+        }
+        // Sort low to high PCI device
+        if (left->pci_device < right->pci_device) {
+            return -1;
+        } else if (left->pci_device > right->pci_device) {
+            return 1;
+        }
+        // Sort low to high PCI function
+        if (left->pci_function < right->pci_function) {
+            return -1;
+        } else if (left->pci_function > right->pci_function) {
+            return 1;
+        }
+    }
+
+    // Somehow we have a tie above, so XOR vendorID and deviceID and compare
+    uint32_t left_xord_dev_vend = left->device_id ^ left->vendor_id;
+    uint32_t right_xord_dev_vend = right->device_id ^ right->vendor_id;
+    if (left_xord_dev_vend < right_xord_dev_vend) {
+        return -1;
+    } else if (right_xord_dev_vend < left_xord_dev_vend) {
+        return 1;
+    }
+    return 0;
+}
+
+// Search for the default device using the loader environment variable and then the Mesa
+// environment variable.  Both are formated the same.  This way, we can start working in
+// the same fashion and the layer can "migrate" away over time.
+const char loader_device_select_env_var[2][32] = {"VK_LOADER_DEVICE_SELECT", "MESA_VK_DEVICE_SELECT"};
```

Comment:
This seems to be undocumented behavior. I like the direction, I just don't know of the cleanlyness of shipping loader source that actively uses the same env-var.

Reply:
I was just using it since it will override our order any way.  I could document it or remove it.  What do you prefer?

Reply:
Never mind, I'll just get rid of it.  It's the only mesa thing left.